File: Programmers/codingTest_Kit/DFS_BFS/43164.cpp
#include <string>
#include <vector>
#include <algorithm>
#define MAX 10001

using namespace std;

bool visited[MAX];

bool DFS(string current, vector<vector<string>> &tickets, vector<string> &answer, int ticketIndex){
    answer.push_back(current);
    if(ticketIndex == tickets.size()) return true;

    for(int i = 0; i < tickets.size(); i++){
        if(tickets[i][0] == current && !visited[i]){
            visited[i] = true;
            if(DFS(tickets[i][1], tickets, answer, ticketIndex+1)){
                return true;
            }
            visited[i] = false;
        }
    }
    answer.pop_back();
    return false;
}

vector<string> solution(vector<vector<string>> tickets) {
    vector<string> answer;
    sort(tickets.begin(), tickets.end());
    DFS("ICN", tickets, answer, 0);
    return answer;
}

File: Programmers/codingTest_Kit/DFS_BFS/43164.py
from collections import defaultdict
def solution(tickets):
    routes = init_graph(tickets)
    for r in routes:
        routes[r].sort()

    N = len(tickets)
    answer = dfs("ICN", ["ICN"], routes, N)

    return answer

def init_graph(tickets):
    routes = defaultdict(list)
    for key, value in tickets:
        routes[key].append(value)
    return routes

def dfs(key, footprint, routes, N):
    if len(footprint) == N + 1:
        return footprint

    for idx, country in enumerate(routes[key]):
        routes[key].pop(idx)

        fp = footprint[:]
        fp.append(country)

        ret = dfs(country, fp, routes, N)
        if ret: return ret

        routes[key].insert(idx, country)